path/spanning_tree.go

```diff
@@ -5,101 +5,173 @@
+		indexOf: make(map[int]int, len(nodes)-1),
+		nodes:   make([]concrete.Edge, 0, len(nodes)-1),
+	}
+	for _, u := range nodes[1:] {
+		heap.Push(q, concrete.Edge{F: u, W: math.Inf(1)})
 	}
 
-	dst.AddNode(nlist[0])
-	remain := make(internal.IntSet)
-	for _, node := range nlist[1:] {
-		remain.Add(node.ID())
+	u := nodes[0]
+	for _, v := range g.From(u) {
+		w, ok := g.Weight(u, v)
+		if !ok {
+			panic("prim: unexpected invalid weight")
+		}
+		q.update(v, u, w)
 	}
 
-	edgeList := g.Edges()
-	for remain.Count() != 0 {
-		var edges []concrete.Edge
-		for _, e := range edgeList {
-			u := e.From()
-			v := e.To()
-			if (dst.Has(u) && remain.Has(v.ID())) || (dst.Has(v) && remain.Has(u.ID())) {
-				w, ok := weight(u, v)
+	var w float64
+	for q.Len() > 0 {
+		e := heap.Pop(q).(concrete.Edge)
+		if e.To() != nil && g.HasEdge(e.From(), e.To()) {
+			dst.SetEdge(e)
+			w += e.Weight()
+		}
+
+		u = e.From()
+		for _, n := range g.From(u) {
+			if key, ok := q.key(n); ok {
+				w, ok := g.Weight(u, n)
 				if !ok {
 					panic("prim: unexpected invalid weight")
 				}
-				edges = append(edges, concrete.Edge{F: u, T: v, W: w})
+				if w < key {
+					q.update(n, u, w)
+				}
 			}
 		}
-
-		sort.Sort(byWeight(edges))
-		min := edges[0]
-		dst.SetEdge(min)
-		remain.Remove(min.From().ID())
 	}
+	return w
+}
+
+// primQueue is a Prim's priority queue. The priority queue is a
+// queue of edge From nodes keyed on the minimum edge weight to
+// a node in the set of nodes already connected to the minimum
+// spanning forest.
+type primQueue struct {
+	indexOf map[int]int
+	nodes   []concrete.Edge
+}
+
+func (q *primQueue) Less(i, j int) bool {
+	return q.nodes[i].Weight() < q.nodes[j].Weight()
+}
 
+func (q *primQueue) Swap(i, j int) {
+	q.indexOf[q.nodes[i].From().ID()] = j
+	q.indexOf[q.nodes[j].From().ID()] = i
+	q.nodes[i], q.nodes[j] = q.nodes[j], q.nodes[i]
 }
 
-// Kruskal generates a minimum spanning tree of g by greedy tree coalesence, placing
-// the result in the destination. The destination is not cleared first.
-func Kruskal(dst graph.MutableUndirected, g EdgeListerGraph) {
-	var weight Weighting
-	if wg, ok := g.(graph.Weighter); ok {
-		weight = wg.Weight
-	} else {
-		weight = UniformCost(g)
+func (q *primQueue) Len() int {
+	return len(q.nodes)
+}
+
+func (q *primQueue) Push(x interface{}) {
+	n := x.(concrete.Edge)
+	q.indexOf[n.From().ID()] = len(q.nodes)
+	q.nodes = append(q.nodes, n)
+}
+
+func (q *primQueue) Pop() interface{} {
+	n := q.nodes[len(q.nodes)-1]
+	q.nodes = q.nodes[:len(q.nodes)-1]
+	delete(q.indexOf, n.From().ID())
+	return n
+}
+
+// key returns the key for the node u and whether the node is
+// in the queue. If the node is not in the queue, key is returned
+// as +Inf.
+func (q *primQueue) key(u graph.Node) (key float64, ok bool) {
+	i, ok := q.indexOf[u.ID()]
+	if !ok {
+		return math.Inf(1), false
+	}
+	return q.nodes[i].Weight(), ok
+}
+
+// update updates the node's position in the queue with the new key.
```

Comment:
May I have one last request to update this comment to say which node's position is updated (because two are given) and what is the role of the other one?